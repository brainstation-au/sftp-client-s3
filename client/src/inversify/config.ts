import { Container, interfaces } from 'inversify';

export const containerOptions: interfaces.ContainerOptions = {
  autoBindInjectable: true,
  skipBaseClassChecks: true,
  defaultScope: 'Singleton',
};

export const container = new Container(containerOptions);
