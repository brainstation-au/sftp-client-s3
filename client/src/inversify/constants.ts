export const S3_TO_SERVER_OPTIONS = Symbol.for('s3-to-server');
export const SERVER_TO_S3_OPTIONS = Symbol.for('server-to-s3');
