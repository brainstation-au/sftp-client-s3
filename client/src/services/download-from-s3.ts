import { S3 } from 'aws-sdk';
import * as fs from 'fs';

export const downloadFromS3 = async (bucket: string, key: string, localPath: string): Promise<void> => {
  const s3 = new S3({apiVersion: '2006-03-01'});
  const request = s3.getObject({
    Bucket: bucket,
    Key: key,
  });
  const w = fs.createWriteStream(localPath);

  return new Promise((resolve, reject) => {
    request.createReadStream()
      .pipe(w)
      .on('close', resolve)
      .on('error', reject);
  });
};
