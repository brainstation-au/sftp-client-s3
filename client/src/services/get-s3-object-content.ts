import { S3 } from 'aws-sdk';
import { parseS3Uri } from './parse-s3-uri';

export const getS3ObjectContent = async (uri: string): Promise<string> => {
  const {bucket, key} = parseS3Uri(uri);
  console.log(bucket,key);

  const s3 = new S3({apiVersion: '2006-03-01'});
  const data = await s3.getObject({
    Bucket: bucket,
    Key: key,
  }).promise();

  return (data.Body as Buffer).toString('utf-8');
};
