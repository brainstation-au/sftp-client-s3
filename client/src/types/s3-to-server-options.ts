import { Boolean, Partial, Record, Static, String } from 'runtypes';
import { ServerParams } from './server-params';

export const S3ToServerOptions = Record({
  bucket: String,
  s3Key: String,
  override: Boolean,
})
  .And(ServerParams)
  .And(Partial({
    gzip: Boolean,
  }));

// eslint-disable-next-line @typescript-eslint/no-redeclare
export type S3ToServerOptions = Static<typeof S3ToServerOptions>;
