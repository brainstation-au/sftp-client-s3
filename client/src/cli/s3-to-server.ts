import { Arguments, Argv, MiddlewareFunction } from 'yargs';
import { container } from '../inversify/config';
import { S3_TO_SERVER_OPTIONS } from '../inversify/constants';
import { getS3ObjectContent } from '../services/get-s3-object-content';
import { s3ToServer } from '../services/s3-to-server';
import { S3ToServerOptions } from '../types/s3-to-server-options';
import { Arguments as S3ToServerArguments } from './s3-to-server-arguments';

export const command = 's3-to-server';

export const description = 'Get a file from S3 and upload that on the SFTP server';

export const builder = (yargs: Argv<unknown>): Argv<Partial<S3ToServerArguments>> => yargs
  .option('host', {
    alias: ['sftp-host', 'h'],
    default: process.env['SFTP_HOST'],
    demandOption: !('SFTP_HOST' in process.env),
    description: 'SFTP host IP address or URL',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('port', {
    alias: ['sftp-port', 'p'],
    default: +(process.env['SFTP_PORT'] || 22),
    description: 'SFTP host port number',
    nargs: 1,
    type: 'number',
  })
  .option('user', {
    alias: ['sftp-user', 'username', 'u'],
    default: process.env['SFTP_USER'],
    demandOption: !('SFTP_USER' in process.env),
    description: 'SFTP username',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('private-key-s3-uri', {
    default: process.env['PRIVATE_KEY_S3_URI'],
    demandOption: !('PRIVATE_KEY_S3_URI' in process.env),
    description: 'S3 URI for the private key to authenticate SFTP session',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('location', {
    alias: ['remote-location', 'l'],
    default: process.env['REMOTE_LOCATION'],
    demandOption: !('REMOTE_LOCATION' in process.env),
    description: 'Path to the file location in SFTP server',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('bucket', {
    alias: ['bucket-name', 'b'],
    default: process.env['BUCKET_NAME'],
    demandOption: !('BUCKET_NAME' in process.env),
    description: 'S3 bucket name',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('s3-key', {
    default: process.env['S3_KEY'],
    demandOption: !('S3_KEY' in process.env),
    description: 'S3 key for the file to upload',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('override', {
    default: process.env['OVERRIDE'] === 'true' || false,
    description: 'Override a file in SFTP server if already exists',
    type: 'boolean',
  })
  .option('gzip', {
    alias: ['compress'],
    default: process.env['COMPRESS'] === 'true' || false,
    description: 'Compress file content if not already compressed',
    type: 'boolean',
  })
  .option('encrypt', {
    alias: ['e'],
    default: process.env['ENCRYPT'] === 'true' || false,
    description: 'Encrypt file content with GPG public key',
    type: 'boolean',
    implies: ['gpg-public-key-s3-uri'],
  })
  .option('gpg-public-key-s3-uri', {
    default: process.env['GPG_PUBLIC_KEY_S3_URI'],
    description: 'S3 URI of the GPG public key to encrypt file content',
    nargs: 1,
    type: 'string',
  });

const downloadPrivateKey: MiddlewareFunction<Partial<S3ToServerArguments>> = async (argv) => {
  const privateKey = argv.privateKeyS3Uri && await getS3ObjectContent(argv.privateKeyS3Uri);
  return {...argv, privateKey};
};

const downloadGpgPublicKey: MiddlewareFunction<Partial<S3ToServerArguments>> = async (argv) => {
  const gpgPublicKey = argv.gpgPublicKeyS3Uri && await getS3ObjectContent(argv.gpgPublicKeyS3Uri);
  return {...argv, gpgPublicKey};
};

export const middlewares = [
  downloadPrivateKey,
  downloadGpgPublicKey,
];

export const handler = async (argv: Arguments<Partial<S3ToServerArguments>>): Promise<void> => {
  const options = S3ToServerOptions.check(argv);
  container.bind<S3ToServerOptions>(S3_TO_SERVER_OPTIONS).toConstantValue(options);
  return s3ToServer(options);
};
