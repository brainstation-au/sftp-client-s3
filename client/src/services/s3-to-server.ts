import * as fs from 'fs';
import * as path from 'path';
import * as zlib from 'zlib';
import { S3ToServerOptions } from '../types/s3-to-server-options';
import { downloadFromS3 } from './download-from-s3';
import { localStorageLocation } from './local-storage-location';
import { encrypt } from './openpgp';
import { uploadToSftpServer } from './upload-to-sftp-server';

export const s3ToServer = async (options: S3ToServerOptions): Promise<string> => {
  const { bucket, s3Key } = options;
  const filename = path.basename(s3Key);
  const isGzipped = path.extname(filename).toLowerCase() === '.gz';
  const localDir = localStorageLocation();
  const localPath = path.join(localDir, filename);

  await downloadFromS3(bucket, s3Key, localPath);
  console.log(`s3://${bucket}/${s3Key} has been downloaded as ${localPath}`);

  if (options.encrypt) {
    const fileContent: Buffer = fs.readFileSync(localPath);
    const clearContent: Buffer = isGzipped ? zlib.gunzipSync(fileContent) : fileContent;
    const encClearContent: string = await encrypt(clearContent.toString(), options.gpgPublicKey);
    const encContent: Buffer = isGzipped ? zlib.gzipSync(Buffer.from(encClearContent)) : Buffer.from(encClearContent);
    fs.writeFileSync(localPath, encContent);
  }

  let uploadPath = localPath;
  if (!isGzipped && options.gzip) {
    uploadPath = localPath + '.gz';
    fs.writeFileSync(uploadPath, zlib.gzipSync(fs.readFileSync(localPath)));
  }

  return uploadToSftpServer(options, uploadPath);
};
