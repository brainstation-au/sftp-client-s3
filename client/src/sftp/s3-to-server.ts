import { S3ToServerOptions } from '../types/s3-to-server-options';

export const s3ToServer = async (options: S3ToServerOptions): Promise<void> => {
  console.log(options);
};
