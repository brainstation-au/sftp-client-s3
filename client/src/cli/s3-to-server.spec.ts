import yargs, { Arguments } from 'yargs';
import * as s3ToServer from './s3-to-server';
import * as handler from '../services/s3-to-server';
import * as s3Content from '../services/get-s3-object-content';
import { container } from '../inversify/config';
import { S3ToServerOptions } from '../types/s3-to-server-options';
import { SERVER_PARAMS } from '../inversify/constants';
jest.mock('../services/s3-to-server');
jest.mock('../services/get-s3-object-content');
const mockedHandler = handler as jest.Mocked<typeof handler>;
const mockedS3Content = s3Content as jest.Mocked<typeof s3Content>;

describe('s3-to-server', () => {
  const parser = yargs.command(s3ToServer).help();
  const parseArgs = (args: string): Promise<string> => new Promise((resolve, reject) => {
    parser.parse(args, (err: Error | undefined, _argv: Arguments, output: string): void => {
      if (err) reject(err);
      resolve(output);
    });
  });

  describe('help', () => {
    test('basic help output on parent', async () => {
      await expect(parseArgs('--help')).resolves.toMatchSnapshot();
    });

    test('basic help output on command', async () => {
      await expect(parseArgs('s3-to-server --help')).resolves.toMatchSnapshot();
    });
  });

  describe('command options', () => {
    describe('required params are not available', () => {
      test('with no arguments', async () => {
        await expect(parseArgs('s3-to-server')).rejects
          .toThrow('Missing required arguments: host, user, private-key-s3-uri, location, bucket, s3-key');
      });

      test('with few arguments', async () => {
        await expect(parseArgs('s3-to-server --host sftphost --location /download/from/here -b test-bucket')).rejects
          .toThrow('Missing required arguments: user, private-key-s3-uri, s3-key');
      });
    });

    describe('required params are available', () => {
      let output: string;

      beforeAll(async () => {
        mockedS3Content.getS3ObjectContent.mockResolvedValue('secret-code');
        process.env['SFTP_HOST'] = 'sftphost';
        process.env['SFTP_USER'] = 'test_user';
        process.env['S3_KEY'] = 'upload/here/filename.txt';
        process.env['FILENAME'] = 'something.txt';

        output = await parseArgs(
          's3-to-server --port 22 --private-key-s3-uri s3-uri --location /download/from/here -b test-bucket --gzip'
        );
      });

      afterAll(() => {
        mockedHandler.s3ToServer.mockReset();
        mockedS3Content.getS3ObjectContent.mockReset();
        delete process.env['SFTP_HOST'];
        delete process.env['SFTP_USER'];
        delete process.env['S3_KEY'];
        delete process.env['FILENAME'];
      });

      test('resolves successfully', () => {
        expect(output).toEqual('');
      });

      test('options are available in global container', () => {
        expect(container.get<S3ToServerOptions>(SERVER_PARAMS)).toEqual(expect.objectContaining({
          host: 'sftphost',
          port: 22,
          username: 'test_user',
          privateKey: 'secret-code',
          location: '/download/from/here',
          filename: 'something.txt',
        }));
      });

      test('calls handler with arguments', async () => {
        expect(mockedHandler.s3ToServer).toHaveBeenCalledTimes(1);
        expect(mockedHandler.s3ToServer).toHaveBeenCalledWith(expect.objectContaining({
          host: 'sftphost',
          port: 22,
          user: 'test_user',
          privateKey: 'secret-code',
          location: '/download/from/here',
          bucket: 'test-bucket',
          s3Key: 'upload/here/filename.txt',
          gzip: true,
          rm: false,
          filename: 'something.txt',
        }));
      });
    });
  });
});
