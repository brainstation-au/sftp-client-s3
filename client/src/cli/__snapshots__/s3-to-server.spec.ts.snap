// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`s3-to-server help basic help output on command 1`] = `
"jest s3-to-server

Get a file from S3 and upload that on the SFTP server

Options:
      --version                        Show version number             [boolean]
      --help                           Show help                       [boolean]
  -h, --host, --sftp-host              SFTP host IP address or URL
                                                             [string] [required]
  -p, --port, --sftp-port              SFTP host port number
                                                          [number] [default: 22]
  -u, --user, --sftp-user, --username  SFTP username         [string] [required]
      --private-key-s3-uri             S3 URI for the private key to
                                       authenticate SFTP session
                                                             [string] [required]
  -l, --location, --remote-location    Path to the file location in SFTP server
                                                             [string] [required]
  -b, --bucket, --bucket-name          S3 bucket name        [string] [required]
      --s3-key                         S3 key for the file to upload
                                                             [string] [required]
      --override                       Override a file in SFTP server if already
                                       exists         [boolean] [default: false]
      --gzip, --compress               Compress file content if not already
                                       compressed     [boolean] [default: false]"
`;

exports[`s3-to-server help basic help output on parent 1`] = `
"jest [command]

Commands:
  jest s3-to-server  Get a file from S3 and upload that on the SFTP server

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]"
`;
